cmd/generator/main.go

```diff
@@ -1,27 +1,78 @@
-/*
-Copyright 2021 Upbound Inc.
-*/
+// Copyright 2022 Upbound Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
 
 package main
 
 import (
+	"encoding/json"
 	"fmt"
 	"os"
 	"path/filepath"
-
-	"github.com/upbound/upjet/pkg/pipeline"
+	"sort"
+	"strings"
 
 	"github.com/crossplane-contrib/provider-confluent/config"
+	ujconfig "github.com/crossplane/upjet/pkg/config"
+	"github.com/crossplane/upjet/pkg/pipeline"
+	"gopkg.in/alecthomas/kingpin.v2"
 )
 
 func main() {
-	if len(os.Args) < 2 || os.Args[1] == "" {
-		panic("root directory is required to be given as argument")
+	var (
+		app                   = kingpin.New("generator", "Run Upjet code generation pipelines for provider-gcp").DefaultEnvars()
+		repoRoot              = app.Arg("repo-root", "Root directory for the provider repository").Required().String()
+		skippedResourcesCSV   = app.Flag("skipped-resources-csv", "File path where a list of skipped (not-generated) Terraform resource names will be stored as a CSV").Envar("SKIPPED_RESOURCES_CSV").String()
+		generatedResourceList = app.Flag("generated-resource-list", "File path where a list of the generated resources will be stored.").Envar("GENERATED_RESOURCE_LIST").Default("../config/generated.lst").String()
+	)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	absRootDir, err := filepath.Abs(*repoRoot)
+	if err != nil {
+		panic(fmt.Sprintf("cannot calculate the absolute path with %s", *repoRoot))
+	}
+	p := config.GetProvider()
+	dumpGeneratedResourceList(p, generatedResourceList)
+	dumpSkippedResourcesCSV(p, skippedResourcesCSV)
+	pipeline.Run(p, absRootDir)
+}
+
+func dumpGeneratedResourceList(p *ujconfig.Provider, targetPath *string) {
+	if len(*targetPath) == 0 {
+		return
 	}
-	rootDir := os.Args[1]
-	absRootDir, err := filepath.Abs(rootDir)
+	generatedResources := make([]string, 0, len(p.Resources))
+	for name := range p.Resources {
+		generatedResources = append(generatedResources, name)
+	}
+	sort.Strings(generatedResources)
+	buff, err := json.Marshal(generatedResources)
 	if err != nil {
-		panic(fmt.Sprintf("cannot calculate the absolute path with %s", rootDir))
+		panic(fmt.Sprintf("Cannot marshal native schema versions to JSON: %s", err.Error()))
```

Comment:
Would it be better to return these errors and deal with them in the main function?

Reply:
This is a good question. To be honest, I copied this code from the official Upbound GCP Provider: https://github.com/upbound/provider-gcp/blob/main/cmd/generator/main.go

In general though, since this function is directly called from the main function / entrypoint and the error is unrecoverable and should be considered immediately fatal, then panic is fine.